Azure Service Discovery: "panic: runtime error: invalid memory address or nil pointer dereference" with Managed Identity

<!--

    Please do *NOT* ask support questions in Github issues.

    If your issue is not a feature request or bug report use our
    community support.

    https://prometheus.io/community/

    There is also commercial support available.

    https://prometheus.io/support-training/

-->

**What did you do?**
We have prometheus getting windows_exporter metrics in Azure
This is the Design
![image](https://user-images.githubusercontent.com/8511684/134504943-286f3f27-bea5-4b55-9c56-b24a76720265.png)





**What did you expect to see?**
We have tested prometheus Azure Discovery setup with Managed Identity with Windows_Exporter_VM_Resource_Group  in  the same azure subscription than the Monitoring_Resource_Group and everything work fine

**NOTE** : The **Windows_Exporter_VM_Resource_Group** and **Monitoring_Resource_Group** are resource groups names of the Design above. Just to be consistent. 

Prometheus logs

```
level=info ts=2021-09-23T12:27:32.414Z caller=main.go:979 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
level=info ts=2021-09-23T12:27:32.420Z caller=main.go:1016 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml totalDuration=6.4103ms db_storage=900ns remote_storage=1.8µs web_handler=600ns query_engine=1.3µs scrape=486.8µs scrape_sd=232.4µs notify=1.9µs notify_sd=4.2µs rules=2.1µs
level=info ts=2021-09-23T12:27:32.420Z caller=main.go:794 msg="Server is ready to receive web requests."
```

and the Prometheus target pages is
![image](https://user-images.githubusercontent.com/8511684/134506830-f38551f3-b6d5-45ce-b09e-600afcdb510c.png)

NOTE: The states of the azure_nodes_drives are Down since the VNET peering is not set yet but the service discovery is working fine :-)


**What did you see instead? Under which circumstances?**
If the Windows_Exporter_VM_Resource_Group is in **different** azure subscription than the Monitoring_Resource_Group we get the following error in prometheus logs

```
level=info ts=2021-09-23T12:08:35.951Z caller=main.go:794 msg="Server is ready to receive web requests."
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xe8d019]

goroutine 864 [running]:
github.com/prometheus/prometheus/discovery/azure.mapFromVM({{0x0}, 0x0, 0xc00079dc80, 0xc001739e00, 0x0, 0x0, 0xc00173d0a0, 0xc00173d090, 0xc00173ce80, 0xc00173ce90, ...})
        /app/discovery/azure/azure.go:468 +0xf9
github.com/prometheus/prometheus/discovery/azure.(*azureClient).getVMs(0xc00147c470, {0x34448d0, 0xc0005dcac0})
        /app/discovery/azure/azure.go:397 +0x265
github.com/prometheus/prometheus/discovery/azure.(*Discovery).refresh(0xc000bd5770, {0x34448d0, 0xc0005dcac0})
        /app/discovery/azure/azure.go:266 +0x2db
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).refresh(0xc0005dc9c0, {0x34448d0, 0xc0005dcac0})
        /app/discovery/refresh/refresh.go:117 +0x139
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).Run(0xc0005dc9c0, {0x34448d0, 0xc0005dcac0}, 0xc00063c300)
        /app/discovery/refresh/refresh.go:76 +0x77
created by github.com/prometheus/prometheus/discovery.(*Manager).startProvider
        /app/discovery/manager.go:201 +0x3b3
        
```

The Prometheus target page is 
![image](https://user-images.githubusercontent.com/8511684/134506290-4ec6bc75-4014-450c-9cd1-d65ef24f0c02.png)

**Environment**

* System information:

HOST: Windows 10 Enterprise Version 20H2
![image](https://user-images.githubusercontent.com/8511684/134507518-ad9f6e4c-80ae-49bc-814c-c7a275a76005.png)


* Prometheus version:
Prometheus Docker image version: 2.29.2 and 2.30.0
	![image](https://user-images.githubusercontent.com/8511684/134507958-793231e4-4f71-4ee0-a043-7e352d01bdd5.png)

* Windows Exporter version:

	Windows_exporter version: 0.16.0

* Docker Desktop with WSL 2 backend:
![image](https://user-images.githubusercontent.com/8511684/134509147-c36414e7-b903-43e9-a5cf-21b7d5084bee.png)

* Docker Compose config:

```
version: '3.7'

services:

  prometheus:
    image: prom/prometheus:v2.30.0
    container_name: prometheus
    extra_hosts: ["host.docker.internal:host-gateway"]
    restart: always
    volumes:
      - C:/Profana/Prometheus/config/:/etc/prometheus/
      - C:/Profana/Prometheus/data:/prometheus
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--storage.tsdb.retention.size=20GB'
      - '--storage.tsdb.retention.time=120d'

    ports:
      - 9090:9090
      - 80:80

  grafana:
    image: grafana/grafana:8.1.2
    container_name: grafana
    restart: always
    depends_on:
      - prometheus
    ports:
      - 3000:3000
    environment:
      - GF_LOG_MODE=console file
    volumes:
      - C:/Profana/Grafana/data/:/var/lib/grafana
      - C:/Profana/Grafana/logs/:/var/log/grafana
      - C:/Profana/Grafana/plugins/:/var/lib/grafana/plugins
 ```

* Prometheus configuration file:
```
# my global config
global:
  scrape_interval:     60s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 60s # Evaluate rules every 15 seconds. The default is every 1 minute.

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"
scrape_configs:
  - job_name: 'prometheus'
    static_configs:
    - targets: ['localhost:9090']

  - job_name: 'azure_nodes_drives'
    azure_sd_configs:
      - environment: AzurePublicCloud
        authentication_method: ManagedIdentity
        subscription_id: 'xxxxxxxxxxxxxxx'
        port: 9182
        refresh_interval: 60s
    relabel_configs:
      - source_labels: [__meta_azure_machine_resource_group]
        regex: (.+)
        action: keep
      - source_labels: [__meta_azure_machine_computer_name]
        target_label: instance
      - source_labels: [__meta_azure_machine_private_ip]
        regex: (.+)
        replacement: ${1}:9182
        target_label: __address__
```


* Logs:
Not working case
```
level=info ts=2021-09-23T12:08:35.951Z caller=main.go:794 msg="Server is ready to receive web requests."
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xe8d019]

goroutine 864 [running]:
github.com/prometheus/prometheus/discovery/azure.mapFromVM({{0x0}, 0x0, 0xc00079dc80, 0xc001739e00, 0x0, 0x0, 0xc00173d0a0, 0xc00173d090, 0xc00173ce80, 0xc00173ce90, ...})
        /app/discovery/azure/azure.go:468 +0xf9
github.com/prometheus/prometheus/discovery/azure.(*azureClient).getVMs(0xc00147c470, {0x34448d0, 0xc0005dcac0})
        /app/discovery/azure/azure.go:397 +0x265
github.com/prometheus/prometheus/discovery/azure.(*Discovery).refresh(0xc000bd5770, {0x34448d0, 0xc0005dcac0})
        /app/discovery/azure/azure.go:266 +0x2db
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).refresh(0xc0005dc9c0, {0x34448d0, 0xc0005dcac0})
        /app/discovery/refresh/refresh.go:117 +0x139
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).Run(0xc0005dc9c0, {0x34448d0, 0xc0005dcac0}, 0xc00063c300)
        /app/discovery/refresh/refresh.go:76 +0x77
created by github.com/prometheus/prometheus/discovery.(*Manager).startProvider
        /app/discovery/manager.go:201 +0x3b3
```

--- a/discovery/azure/azure.go
+++ b/discovery/azure/azure.go
@@ -464,7 +464,9 @@ func mapFromVM(vm compute.VirtualMachine) virtualMachine {
 		}
 	}
 
-	if vm.VirtualMachineProperties != nil && vm.VirtualMachineProperties.OsProfile != nil {
+	if vm.VirtualMachineProperties != nil &&
+		vm.VirtualMachineProperties.OsProfile != nil &&
+		vm.VirtualMachineProperties.OsProfile.ComputerName != nil {
 		computerName = *(vm.VirtualMachineProperties.OsProfile.ComputerName)
 	}
 

